python/feldera/rest/client.py

```diff
@@ -0,0 +1,455 @@
+            time.sleep(0.1)
+
+    def shutdown_pipeline(self, pipeline_name: str):
+        """
+        Shutdown a pipeline
+        :param pipeline_name: The name of the pipeline to shutdown
+        """
+        self.http.post(
+            path=f"/pipelines/{pipeline_name}/shutdown",
+        )
+
+        while True:
+            status = self.get_pipeline(pipeline_name).state.get("current_status")
+
+            if status == "Shutdown":
+                break
+            elif status == "Failed":
+                raise RuntimeError(f"Failed to shutdown pipeline")
+
+            logging.debug("still shutting down %s, waiting for 100 more milliseconds", pipeline_name)
+            time.sleep(0.1)
+
+    # TODO: better name for this method
+    def push_to_pipeline(
+            self,
+            pipeline_name: str,
+            table_name: str,
+            format: str,
+            data: list[list | str | dict],
+            array: bool = False,
+            force: bool = False,
+            update_format: str = "raw",
+    ):
+        """
+        Insert data into a pipeline
+        :param pipeline_name: The name of the pipeline
+        :param table_name: The name of the table
+        :param format: The format of the data, either "json" or "csv"
+        :param array: Set True if updates in this stream are packed into JSON arrays, used in conjunction with the
+            "json" format
+        :param force: If True, the data will be inserted even if the pipeline is paused
+        :param update_format: JSON data change event format, used in conjunction with the "json" format,
+        the default value is "insert_delete", other supported formats: "weighted", "debezium", "snowflake", "raw"
+        :param data: The data to insert
+        """
+
+        if format not in ["json", "csv"]:
+            raise ValueError("format must be either 'json' or 'csv'")
+
+        if update_format not in ["insert_delete", "weighted", "debezium", "snowflake", "raw"]:
+            raise ValueError("update_format must be one of 'insert_delete', 'weighted', 'debezium', 'snowflake', 'raw'")
+
+        # python sends `True` which isn't accepted by the backend
+        array = "true" if array else "false"
+        force = "true" if force else "false"
+
+        params = {
+            "force": force,
+            "format": format,
+        }
+
+        if format == "json":
+            params["array"] = array
+            params["update_format"] = update_format
+
+        content_type = "application/json"
+
+        self.start_pipeline(pipeline_name)
+
+        if format == "csv":
+            content_type = "text/csv"
+            data = bytes(str(data), "utf-8")
+
+        self.http.post(
+            path=f"/pipelines/{pipeline_name}/ingress/{table_name}",
+            params=params,
+            content_type=content_type,
+            body=data,
+        )
+
+    def listen_to_pipeline(

```

Comment:
Hmm, the current API is not fully implemented for snapshots in particular. Maybe longer term it would make sense to have dedicated functions like `watch(table_or_view)`,  `snapshot(table_or_view)`, `quantiles(table_or_view)`, `neighborhood(table_or_view, ...)`. Not this PR though, but in future.